src/snowflake/snowpark/modin/plugin/extensions/dataframe_overrides.py

```diff
@@ -459,104 +464,222 @@ def __init__(
+                    # The list has the shape (3, 3) while the ndarray has the shape (3,).
+                    # If only some data is a Snowpark pandas object, convert it to pandas objects.
+                    res = []
+                    for v in data:
+                        if isinstance(v, (Index, BasePandasDataset)):
+                            res.append(v.to_pandas())
+                        else:
+                            # # Need to convert this is a native pandas object since native pandas incorrectly
+                            # # tries to perform `get_indexer` on it. Specify dtype=object so that pandas does not
+                            # # cast the data provided. In some cases, None turns to NaN, which is not desired.
+                            # res.append(native_pd.Index(v, dtype=object) if is_list_like(v) else v)
+                            res.append(v)
+                    data = res
+
+        query_compiler = from_pandas(
+            native_pd.DataFrame(
+                data=data,
+                # Handle setting the index, if it is a lazy index, outside this block.
+                index=None if isinstance(index, (Index, Series)) else index,
+                columns=try_convert_index_to_native(columns),
+                dtype=dtype,
+                copy=copy,
+            )
+        )._query_compiler
+
+    # 2. Setting the index
+    # --------------------
+    # The index is already set if the data is a non-Snowpark pandas object.
+    # If either the data or the index is a Snowpark pandas object, set the index here.
+    if index is not None and (
+        isinstance(index, (Index, Series))
+        or isinstance(data, (Index, BasePandasDataset))
+    ):
+        if isinstance(data, (type(self), Series, type(None))):
+            # The `index` parameter is used to select the rows from `data` that will be in the resultant DataFrame.
+            # If a value in `index` is not present in `data`'s index, it will be filled with a NaN value.
+            # If data is None and an index is provided, set the index.
+            query_compiler = query_compiler.reindex(
+                axis=0, labels=convert_index_to_qc(index)
+            )
+        else:
+            # Performing set index to directly set the index column (joining on row-position instead of index).
+            query_compiler = query_compiler.set_index(
+                convert_index_to_list_of_qcs(index)
+            )
 
-                self._query_compiler = new_qc
-                return
-
-            data = {
-                k: v._to_pandas() if isinstance(v, Series) else v
-                for k, v in data.items()
-            }
-        pandas_df = native_pd.DataFrame(
-            data=try_convert_index_to_native(data),
-            index=try_convert_index_to_native(index),
-            columns=try_convert_index_to_native(columns),
-            dtype=dtype,
-            copy=copy,
+    # 3. If data is a DataFrame, filter result
```

Comment:
you should move all columns operation into Step 1. It is confusing to select columns here again.

Reply:
done!